gemmforge/thread_policies/gemm/dense_sparse.py

```diff
@@ -0,0 +1,64 @@
+from ..abstract_thread_policy import AbstractGemmLikeThreadPolicy, DenseMatrix, SparseMatrix
+from gemmforge.vm import VM
+
+
+class GenericDenseSparseGemmThreadPolicy(AbstractGemmLikeThreadPolicy):
+  def __init__(self,
+               vm: VM,
+               shr_mem_per_op: int,
+               num_threads: int,
+               op1: DenseMatrix,
+               op2: SparseMatrix,
+               res: DenseMatrix):
+    super().__init__(vm, num_threads, op1, op2, res)
+    self._shr_mem_per_op = shr_mem_per_op
+
+  def _estimate_num_registers_per_mult(self, accumulator_length):
+    # Note: derived experimentally
+    factor = self._vm.bytes_per_real() / 4
+    return factor * (32 + accumulator_length)
+
+  def get_num_ops_per_block(self):
+    accumulator_length = self._op2.get_actual_num_cols()
+    max_num_regs_per_thread = self._estimate_num_registers_per_mult(accumulator_length)
+
+    hw_descr = self._vm.get_hw_descr()
+    shr_mem_bytes = self._shr_mem_per_op * self._vm.bytes_per_real()
+    mults_wrt_shr_mem = hw_descr.max_local_mem_size_per_block / shr_mem_bytes
+    mults_wrt_num_regs = hw_descr.max_reg_per_block / (self._num_threads * max_num_regs_per_thread)
+    mults_per_sm = int(min(mults_wrt_shr_mem, mults_wrt_num_regs))
+
+    el_count_if_dense = self._op2.get_actual_num_cols() * self._op2.get_actual_num_rows()
+    el_count_if_sparse = self._op2.get_el_count()
+    factor = (el_count_if_dense / el_count_if_sparse)
+    factor /= 2.0
+    # To prevent using too much shared memory
+    if factor > 1.00/0.15:
```

Comment:
Just as a question—is this empirically chosen or is there some reasoning behind it?

Reply:
It is empirically chosen workaround currently, the both dense thread policy and sparse thread policies may require more than 1024 threads per thread block. Dense needs this if there are more than 1024 rows in matrix A, and dense sparse if the matrices are so sparse (B) that it decides use a lot of matrices. Thus this check. 

To check for threads > 1024? I need to check for the matrix row count here and read from a lexicon how many threads are allowed per thread block (which might change in the future etc. and therefore I decided to go for this limit) According to your suggestions I can change to code as I dont know a better solution.

But I have experimentally seen that multiplying multiple matrices in a thread block doesnt bring significant improvements in any of the considered test cases.

Reply:
And even if it sometimes compiles (e.g., a too sparse matrix b could result multiplying 6 blocks but then there might not be enough shared memory or the kernel uses almost all the available shared memory, then it can be still slower than multiplying less block in some cases), but in a lot of randomly generated cases multiplying 1 or n blocks didnt change too much.